src/dbnode/storage/database.go

```diff
@@ -928,15 +934,112 @@ func (d *db) ReadEncoded(
 			xopentracing.Time("end", end),
 		)
 	}
+
 	defer sp.Finish()
+	return n.ReadEncoded(ctx, id, start, end)
+}
 
+func (d *db) WideQuery(
+	ctx context.Context,
+	namespace ident.ID,
+	query index.Query,
+	queryStart time.Time,
+	shards []uint32,
+	iterOpts index.IterationOptions,
+) ([]ident.IndexChecksum, error) { // FIXME: change when exact type known.
 	n, err := d.namespaceFor(namespace)
 	if err != nil {
 		d.metrics.unknownNamespaceRead.Inc(1)
 		return nil, err
 	}
 
-	return n.ReadEncoded(ctx, id, start, end)
+	var (
+		doneCh = make(chan struct{})
+
+		batchSize = d.opts.WideBatchSize()
+		blockSize = n.Options().IndexOptions().BlockSize()
+
+		collector          = make(chan *ident.IDBatch)
+		collectedChecksums = make([]ident.IndexChecksum, 0, 10)
+		collectorErr       error
+	)
+
+	opts, err := index.NewWideQueryOptions(queryStart, batchSize, blockSize, shards, iterOpts)
+	if err != nil {
+		return nil, err
+	}
+
+	start, end := opts.StartInclusive, opts.EndExclusive
+	ctx, sp, sampled := ctx.StartSampledTraceSpan(tracepoint.DBWideQuery)
+	if sampled {
+		sp.LogFields(
+			opentracinglog.String("wideQuery", query.String()),
+			opentracinglog.String("namespace", namespace.String()),
+			opentracinglog.Int("batchSize", batchSize),
+			xopentracing.Time("start", start),
+			xopentracing.Time("end", end),
+		)
+	}
+
+	defer sp.Finish()
+
+	// Setup consumer
+	go func() {
+		for batch := range collector {
+			if collectorErr != nil {
+				batch.Done()
+				continue
+			}
+
+			var checksum ident.IndexChecksum
+			for i, id := range batch.IDs {
+				checksum, collectorErr = d.fetchIndexChecksum(ctx, n, id, start)
```

Comment:
Don't you need to check if collectorErr != nil before executing this line? Otherwise an error could be in collectorErr and now it gets set to nil on a corresponding execution of this line.

Think it would be cleaner to have an explicit cleanup step where it drains the channel instead of the continue at the top here, otherwise you'll need continues everywhere (like here too).

For instance, perhaps we do this?
```go
	// Setup consumer
	go func() {
		var (
			abortBatch *id.IDBatch
			abortErr error
		)
		defer func() {
			if abortBatch != nil {
				abortBatch.Done()
			}
			if abortErr != nil {
				for batch := range collector {
					batch.Done()
				}
			}
			doneCh <- struct{}{}
		}()

		for batch := range collector {
			var checksum ident.IndexChecksum
			for i, id := range batch.IDs {
				checksum, err := d.fetchIndexChecksum(ctx, n, id, start)
				if err != nil {
					// Set abort vars for cleanup.
					abortBatch, abortErr = batch, err
					return
				}

				// TODO: use index checksum value
				useID := i == len(batch.IDs)-1
				if !useID {
					checksum.ID = checksum.ID[:0]
				}

				collectedChecksums = append(collectedChecksums, checksum)
			}

			batch.Done()
		}
	}()
```

Reply:
Ah yeah, good call; wasn't too focused on getting the error handling in this part working yet, but this will be a good pattern for usage going forwards :+1: